packages/core/src/bin/utils/run.ts

```diff
@@ -93,53 +100,140 @@ export async function run({
+          updateLastEventCheckpoint(indexingService, lastEventCheckpoint);
+      });
+
+    for await (const rawEvents of syncService.getEvents({
+      sources,
+      fromCheckpoint: checkpoint,
+      toCheckpoint: newCheckpoint,
+      limit: 1_000,
+    })) {
+      const events = decodeEvents(indexingService, rawEvents);
+      const result = await processEvents(indexingService, {
+        events,
+      });
+
+      if (result.status === "error") {
+        onReloadableError(result.error);
+        break;
+      } else if (indexingService.isKilled) {
+        break;
+      }
+    }
 
-  const finalizedCheckpoint = await syncService.start();
+    await updateLastEventPromise;
 
-  let isHealthy = false;
-  indexingService.onSerial("eventsProcessed", async ({ toCheckpoint }) => {
-    if (isHealthy) return;
+    checkpoint = newCheckpoint;
 
-    if (isCheckpointGreaterThanOrEqualTo(toCheckpoint, finalizedCheckpoint)) {
+    let isHealthy = false;
+    if (
+      !isHealthy &&
+      isCheckpointGreaterThanOrEqualTo(checkpoint, finalizedCheckpoint)
+    ) {
       isHealthy = true;
       serverService.setIsHealthy(true);
       common.logger.info({ service: "server", msg: "Responding as healthy" });
     }
-  });
 
-  await indexingService.start({
-    indexingFunctions,
-    schema,
-    tableAccess,
-    cachedToCheckpoint,
+    // TODO(kyle) update per network checkpoint
+  };
+
+  const handleReorg = async (safeCheckpoint: Checkpoint) => {
+    // No-op if realtime indexing hasn't started
+    if (isCheckpointGreaterThan(finalizedCheckpoint, checkpoint)) return;
+
+    await indexingStore.revert({
+      checkpoint: safeCheckpoint,
+      isCheckpointSafe: true,
+    });
+    checkpoint = safeCheckpoint;
+    await handleCheckpoint(syncService.checkpoint);
+  };
+
+  type SyncEvent =
+    | {
+        type: "checkpoint";
+        newCheckpoint: Checkpoint;
+      }
+    | {
+        type: "reorg";
+        safeCheckpoint: Checkpoint;
+      };
+
+  const runQueue = createQueue({
+    initialStart: true,
+    browser: false,
+    concurrency: 1,
+    worker: async (syncEvent: SyncEvent) => {
+      switch (syncEvent.type) {
+        case "checkpoint":
+          await handleCheckpoint(syncEvent.newCheckpoint);
+          break;
+
+        case "reorg":
+          await handleReorg(syncEvent.safeCheckpoint);
+          break;
+
+        default:
+          neva(syncEvent);
+      }
+    },
   });
 
-  indexingService.processEvents();
+  syncService.on("checkpoint", async (newCheckpoint) =>
+    runQueue.add({
+      type: "checkpoint",
+      newCheckpoint,
+    }),
+  );
+
+  syncService.on("reorg", async (safeCheckpoint) =>
+    runQueue.add({
+      type: "reorg",
+      safeCheckpoint,
+    }),
+  );
+
+  syncService.on("fatal", onFatalError);
+
+  const finalizedCheckpoint = await syncService.start();
 
   return async () => {
+    runQueue.pause();
+    runQueue.clear();
+    kill(indexingService);
```

Comment:
Overall on this file -  I'm not fully convinced that the indexing "guts" (eg, the queue, the setup event processing, etc) should be exposed here rather than within an indexing service. It seems quite important to test the logic here (eg, what happens when different sequences of checkpoint & reorg events get added to the queue)?